### PR TITLE
Generate Sponsor invoice

### LINKED ISSUE #235: We need to support invoice generation for sponsors
From @andreasneumann 

> Could we add a function to generate a sponsorship invoice?

### PR DESCRIPTION
**Pull request by sonlinux**
Fix #235 

Add support of invoice generation for sponsors.

sponsor_invoice_demo

  
Example Invoice.pdf file as shown above:

[Invoice_qgis-pagkrh.pdf](https://github.com/kartoza/projecta/files/1614451/Invoice_qgis-pagkrh.pdf)

### CODE REVIEW

File: django_project/changes/models/sponsor.py [hunk 1]
```diff
@@ -81,6 +87,12 @@ class Sponsor(models.Model):
         null=True,
         blank=True)
 
+    sponsorship_contact_email = models.CharField(
```
Comment (cchristelis) [inline django_project/changes/models/sponsor.py]: Shouldn't this be an EmailField?

File: django_project/changes/views/sponsor.py [hunk 2]
```diff
@@ -659,3 +663,57 @@ def generate_sponsor_cloud(request, **kwargs):
...
+        context['project_slug'] = project_slug
+        context['sponsor_slug'] = sponsor_slug
+        context['sponsors'] = sponsors
+        if project_slug and sponsor_slug:
+            project = Project.objects.get(slug=project_slug)
+            context['sponsor'] = sponsors.get(
+                project=project,
+                slug=sponsor_slug)
+            context['project'] = project
+            context['invoice_no'] = '12345'
```
Comment (cchristelis) [inline django_project/changes/views/sponsor.py]: this needs to be updated. Perhaps give each model object an ID and use that?

File: django_project/changes/models/sponsor.py (hunk 1 above)
Comment (sonlinux) [reply django_project/changes/models/sponsor.py]: Hi @cchristelis yes it should, I adopted the practise to use CharField as used for the other email field on the same model such as 
``` 
sponsor_email = models.CharField(
        help_text='Input an email of sponsor.',
        max_length=255,
        null=True,
        blank=True)
```

File: django_project/changes/views/sponsor.py (hunk 2 above)
Comment (sonlinux) [reply django_project/changes/views/sponsor.py]: Thanks ill update that shortly.

Comment (timlinux) [reply django_project/changes/views/sponsor.py]: I think we might also need to be able to manually enter an invoice number...

Comment (sonlinux) [reply django_project/changes/views/sponsor.py]: Alright @timlinux ill do just that. 
Thanks

**Comment by sonlinux** [comment]
Hi @timlinux please take a look at a sample downloaded invoice below for review,
Thanks

[Invoice_qgis-pagkrh.pdf](https://github.com/kartoza/projecta/files/1614460/Invoice_qgis-pagkrh.pdf)

**Comment by timlinux** [comment]
Hi @sonlinux Thanks for this! Could you make. Could you also make a little gif showing how the sponsorship contact details are managed?

@andreasneumann do you have any comments?

**Comment by sonlinux** [comment]
Hi @sonlinux Thanks for this! Could you make. Could you also make a little gif showing how the sponsorship contact details are managed?

Alright @timlinux ill do just that.
Thanks

File: django_project/changes/models/sponsor.py (hunk 1 above)
Comment (ann26) [reply django_project/changes/models/sponsor.py]: please add email validators too like this one https://github.com/kartoza/projecta/blob/develop/django_project/certification/models/certifying_organisation.py#L92

File: django_project/changes/models/sponsor.py [hunk 3]
```diff
@@ -49,6 +49,28 @@ class Sponsor(models.Model):
         blank=False,
         unique=False)  # there is a unique together rule in meta class below
 
+    sponsorship_contact_name = models.CharField(
```
Comment (ann26) [inline django_project/changes/models/sponsor.py]: we already have the sponsor contact person name here https://github.com/kartoza/projecta/pull/661/files#diff-371604757162da387f77bda08ad26881L58

File: django_project/changes/models/sponsor.py [hunk 4]
```diff
@@ -49,6 +49,28 @@ class Sponsor(models.Model):
...
+
+    sponsorship_contact_title = models.CharField(
+        _('Sponsorship Representative Title'),
+        max_length=255,
+        null=True,
+        blank=True,
+        help_text=_('Title of the sponsorship representative i.e Treasurer.')
+    )
+
+    sponsorship_office_address = models.TextField(
```
Comment (ann26) [inline django_project/changes/models/sponsor.py]: What's the difference between sponsorship_office_address and address ( https://github.com/kartoza/projecta/pull/661/files#diff-371604757162da387f77bda08ad26881L64 ) in this model?

File: django_project/changes/models/sponsor.py (hunk 1 above)
Comment (ann26) [reply django_project/changes/models/sponsor.py]: Also what's the difference between sponsorship_contact_email and sponsor_email ( https://github.com/kartoza/projecta/pull/661/files#diff-371604757162da387f77bda08ad26881L78 ) in this model?

File: django_project/changes/templates/sponsor/detail.html [hunk 5]
```diff
@@ -19,6 +19,13 @@ <h1>Sponsors (all)</h1>
                    data-title="Delete {{ certifyingorganisation.name }}">
                     <span class="glyphicon glyphicon-minus"></span>
                 </a>
+                {% if user.is_staff or user == project.owner %}
```
Comment (ann26) [inline django_project/changes/templates/sponsor/detail.html]: The sponsorship manager should see this action buttons too.

File: django_project/changes/utils/render_to_pdf.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+
+def links(uri, rel):
+    link = os.path.join(
+        settings.MEDIA_ROOT, uri.replace(settings.MEDIA_URL, ''))
+    return link
+
+
+def render_to_pdf(template_src, context_dict={}):
+    """Generate PDF from parsed template."""
```
Comment (ann26) [inline django_project/changes/utils/render_to_pdf.py]: Please put a blank line after docstring.

**Comment by ann26** [comment]
hi @sonlinux maybe it would be better to redo the migrations because some migrations are just renaming typos? Making it only 1 migrations for the model changes will make it more succinct. Also I see some new fields are redundant?

**Comment by timlinux** [comment]
@sonlinux can you address @ann26 's comments please so we can merge this?

**Comment by sonlinux** [comment]
Alright @timlinux ill do so shortly. 
And thanks @nn26.

**Pull request merged**